From: Colin Whittaker <colin.whittaker@smartrg.com>
Date: Tue 12 Nov 2019 17:59:22 -0800
Subject: Fix MTU in brcmfmac to IEEE80211_MAX_DATA_LEN

Signed-off-by: Colin Whittaker <colin.whittaker@smartrg.com>
--
 core.c |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/core.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/core.c
@@ -880,6 +880,7 @@ struct brcmf_if *brcmf_add_if(struct brc
 		if (!ndev)
 			return ERR_PTR(-ENOMEM);
 
+		ndev->max_mtu = IEEE80211_MAX_DATA_LEN;
 		ndev->needs_free_netdev = true;
 		ifp = netdev_priv(ndev);
 		ifp->ndev = ndev;
