From: Colin Whittaker <colin.whittaker@smartrg.com>
Date: Tue 12 Nov 2019 17:59:22 -0800
Subject: Fix MTU in brcmfmac to IEEE80211_MAX_DATA_LEN

Signed-off-by: Colin Whittaker <colin.whittaker@smartrg.com>
--
 core.c |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/core.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/core.c
@@ -898,6 +898,7 @@ struct brcmf_if *brcmf_add_if(struct brc
 #else
 		netdev_set_def_destructor(ndev);
 #endif
+		ndev->max_mtu = IEEE80211_MAX_DATA_LEN;
 		ifp = netdev_priv(ndev);
 		ifp->ndev = ndev;
 		/* store mapping ifidx to bsscfgidx */
