From: Chad Monroe <chad.monroe@smartrg.com>
Date: Tue 09 Dec 2020 10:46:48 -0700
Subject: brcmfmac: set country code revs for sr400ac

Signed-off-by: Chad Monroe <chad.monroe@smartrg.com>
---
 of.c |   10 ++++++++++
 1 file changed, 10 insertions(+)

--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/of.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/of.c
@@ -20,6 +20,11 @@ static void brcmf_of_probe_cc(struct dev
 		{ "JP", "JP", 78 },
 		{ "US", "Q2", 86 },
 	};
+	static struct brcmfmac_pd_cc_entry smartrg_sr400_cc_ent[] = {
+		{ "US", "US", 0 },
+		{ "CA", "CA", 0 },
+		{ "GB", "GB", 0 },
+	};
 	struct brcmfmac_pd_cc_entry *cc_ent = NULL;
 	int table_size = 0;
 
@@ -28,6 +33,11 @@ static void brcmf_of_probe_cc(struct dev
 		table_size = ARRAY_SIZE(netgear_r8000_cc_ent);
 	}
 
+	if (of_machine_is_compatible("smartrg,sr400ac")) {
+		cc_ent = smartrg_sr400_cc_ent;
+		table_size = ARRAY_SIZE(smartrg_sr400_cc_ent);
+	};
+
 	if (cc_ent && table_size) {
 		struct brcmfmac_pd_cc *cc;
 		size_t memsize;
