Index: backports-5.10.16-1/backport-include/linux/skbuff.h
===================================================================
--- backports-5.10.16-1.orig/backport-include/linux/skbuff.h
+++ backports-5.10.16-1/backport-include/linux/skbuff.h
@@ -386,13 +386,6 @@ static inline struct sk_buff *__skb_peek
 {
 	return list_->next;
 }
-
-#if !LINUX_VERSION_IN_RANGE(4,19,10, 4,20,0)
-static inline void skb_mark_not_on_list(struct sk_buff *skb)
-{
-	skb->next = NULL;
-}
-#endif /* 4.19.10 <= x < 4.20 */
 #endif
 
 #if LINUX_VERSION_IS_LESS(4,11,0)
