--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -2785,6 +2785,12 @@ static int hostapd_ctrl_iface_chan_switc
 		break;
 	}
 
+	if (os_strstr(pos, " auto-ht")) {
+		settings.freq_params.ht_enabled = iface->conf->ieee80211n;
+		settings.freq_params.vht_enabled = iface->conf->ieee80211ac;
+		settings.freq_params.he_enabled = iface->conf->ieee80211ax;
+	}
+
 	if (settings.freq_params.center_freq1)
 		dfs_range += hostapd_is_dfs_overlap(
 			iface, bandwidth, settings.freq_params.center_freq1);
@@ -2822,12 +2828,6 @@ static int hostapd_ctrl_iface_chan_switc
 		return 0;
 	}
 
-	if (os_strstr(pos, " auto-ht")) {
-		settings.freq_params.ht_enabled = iface->conf->ieee80211n;
-		settings.freq_params.vht_enabled = iface->conf->ieee80211ac;
-		settings.freq_params.he_enabled = iface->conf->ieee80211ax;
-	}
-
 	for (i = 0; i < iface->num_bss; i++) {
 
 		/* Save CHAN_SWITCH VHT, HE, and EHT config */
