--- a/src/ap/acs.c
+++ b/src/ap/acs.c
@@ -21,6 +21,7 @@
 #include "ap_config.h"
 #include "hw_features.h"
 #include "acs.h"
+#include <time.h>
 
 /*
  * Automatic Channel Selection
@@ -466,11 +467,19 @@ static int acs_get_bw_center_chan(int fr
 
 static int acs_survey_is_sufficient(struct freq_survey *survey)
 {
+	static time_t last_seconds=0;
+	time_t current_seconds;
+
 	if (!(survey->filled & SURVEY_HAS_NF)) {
 		survey->nf = -95;
-		wpa_printf(MSG_INFO,
-			   "ACS: Survey for freq %d is missing noise floor",
-			   survey->freq);
+
+		current_seconds=time(NULL);
+		if (last_seconds != current_seconds) {
+			wpa_printf(MSG_INFO,
+				   "ACS: Survey for freq %d is missing noise floor",
+				   survey->freq);
+			last_seconds = current_seconds;
+		}
 	}
 
 	if (!(survey->filled & SURVEY_HAS_CHAN_TIME)) {
