--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -81,6 +81,8 @@
 #define HOSTAPD_GLOBAL_CTRL_IFACE_PORT_LIMIT	50
 #endif /* CONFIG_CTRL_IFACE_UDP */
 
+extern enum hostapd_chan_status acs_init(struct hostapd_iface *iface);
+
 static void hostapd_ctrl_iface_send(struct hostapd_data *hapd, int level,
 				    enum wpa_msg_type type,
 				    const char *buf, size_t len);
@@ -1407,6 +1409,23 @@ static int hostapd_ctrl_iface_reload_bss
 }
 
 
+static int hostapd_ctrl_iface_trigger_acs(struct hostapd_iface *iface)
+{
+	enum hostapd_chan_status status;
+
+	iface->manual_acs_triggered = true;
+	// Force this to 1 to get a usable channel with manual ACS
+	iface->conf->ht40_plus_minus_allowed = 1;
+	status = acs_init(iface);
+
+	if (status != HOSTAPD_CHAN_ACS) {
+		wpa_printf(MSG_ERROR, "Triggering of ACS failed (%u)", status);
+		return -1;
+	}
+	return 0;
+}
+
+
 static int hostapd_ctrl_iface_disable(struct hostapd_iface *iface)
 {
 	if (hostapd_disable_iface(iface) < 0) {
@@ -4080,6 +4099,9 @@ static int hostapd_ctrl_iface_receive_pr
 	} else if (os_strcmp(buf, "RELOAD") == 0) {
 		if (hostapd_ctrl_iface_reload(hapd->iface))
 			reply_len = -1;
+	} else if (os_strcmp(buf, "TRIGGER_ACS") == 0) {
+		if (hostapd_ctrl_iface_trigger_acs(hapd->iface))
+			reply_len = -1;
 	} else if (os_strcmp(buf, "DISABLE") == 0) {
 		if (hostapd_ctrl_iface_disable(hapd->iface))
 			reply_len = -1;
--- a/hostapd/hostapd_cli.c
+++ b/hostapd/hostapd_cli.c
@@ -1237,6 +1237,12 @@ static int hostapd_cli_cmd_reload_config
 	return wpa_ctrl_command(ctrl, "RELOAD_CONFIG");
 }
 
+static int hostapd_cli_cmd_trigger_acs(struct wpa_ctrl *ctrl, int argc,
+				      char *argv[])
+{
+	return wpa_ctrl_command(ctrl, "TRIGGER_ACS");
+}
+
 
 static int hostapd_cli_cmd_disable(struct wpa_ctrl *ctrl, int argc,
 				   char *argv[])
@@ -1841,6 +1847,8 @@ static const struct hostapd_cli_cmd host
 	  "<addr> [req_mode=] <measurement request hexdump>  = send a Beacon report request to a station" },
 	{ "reload_wpa_psk", hostapd_cli_cmd_reload_wpa_psk, NULL,
 	  "= reload wpa_psk_file only" },
+	{ "trigger_acs", hostapd_cli_cmd_trigger_acs, NULL,
+	  "= trigger automatic channel selection" },
 #ifdef CONFIG_IEEE80211R_AP
 	{ "reload_rxkhs", hostapd_cli_cmd_reload_rxkhs, NULL,
 	  "= reload R0KHs and R1KHs" },
--- a/hostapd/main.c
+++ b/hostapd/main.c
@@ -386,6 +386,7 @@ hostapd_interface_init(struct hapd_inter
 	}
 
 	iface->interfaces = interfaces;
+	iface->manual_acs_triggered = false;
 
 	for (k = 0; k < debug; k++) {
 		if (iface->bss[0]->conf->logger_stdout_level > 0)
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -700,6 +700,13 @@ struct hostapd_iface {
 	int (*enable_iface_cb)(struct hostapd_iface *iface);
 	int (*disable_iface_cb)(struct hostapd_iface *iface);
 
+	/*
+	 * If ACS is triggered manually via hostapd_cli, then
+	 * we won't be doing channel change at the end of ACS.
+	 * It will be done via chan_switch later on, if needed.
+	 */
+	bool manual_acs_triggered;	
+
 	/* Configured freq of interface is NO_IR */
 	bool is_no_ir;
 };
--- a/src/ap/hw_features.c
+++ b/src/ap/hw_features.c
@@ -1186,6 +1186,33 @@ int hostapd_acs_completed(struct hostapd
 		wpa_msg(iface->bss[0]->msg_ctx, MSG_INFO,
 			ACS_EVENT_COMPLETED "freq=%d channel=%d",
 			iface->freq, iface->conf->channel);
+
+		// Record the best channel for later use.
+		{
+			FILE *pFile = NULL;
+			char filename[32];
+			sprintf (filename, "/tmp/hapd_acs_%s", iface->phy);
+			pFile = fopen(filename, "w");
+			if (!pFile) {
+				wpa_printf(MSG_ERROR, "Unable to open %s for writing", filename);
+				break;
+			}
+			fprintf(pFile, "phy=%s\n", iface->phy);
+			fprintf(pFile, "config_fname=%s\n", iface->config_fname);
+			fprintf(pFile, "freq=%d\n", iface->freq);
+			fprintf(pFile, "channel=%d\n", iface->conf->channel);
+			fclose (pFile);
+		}
+
+		// If ACS was triggered manually via hostapd_cli, then
+		// channel change will happen separately via chan_switch.
+		if (iface->manual_acs_triggered == true) {
+			iface->manual_acs_triggered = false;
+			hostapd_set_state(iface, HAPD_IFACE_ENABLED);
+			wpa_msg(iface->bss[0]->msg_ctx, MSG_INFO, AP_EVENT_ENABLED);
+			return 0;
+		}
+
 		break;
 	case HOSTAPD_CHAN_ACS:
 		wpa_printf(MSG_ERROR, "ACS error - reported complete, but no result available");
--- a/src/ap/acs.c
+++ b/src/ap/acs.c
@@ -308,7 +308,7 @@ static const struct bw_item *bw_desc[] =
 
 static int acs_request_scan(struct hostapd_iface *iface);
 static int acs_survey_is_sufficient(struct freq_survey *survey);
-
+extern void hostapd_ctrl_update_fixup( struct hostapd_config *conf);
 
 static void acs_clean_chan_surveys(struct hostapd_channel_data *chan)
 {
@@ -879,7 +879,7 @@ acs_find_ideal_chan_mode(struct hostapd_
 			continue;
 
 		if (!chan_bw_allowed(chan, bw, secondary_channel != -1, 1)) {
-			wpa_printf(MSG_DEBUG,
+			wpa_printf(MSG_WARNING,
 				   "ACS: Channel %d: BW %u is not supported",
 				   chan->chan, bw);
 			continue;
@@ -892,7 +892,7 @@ acs_find_ideal_chan_mode(struct hostapd_
 		      iface->conf->secondary_channel) ||
 		     is_6ghz_freq(chan->freq)) &&
 		    !acs_usable_bw_chan(chan, ACS_BW40)) {
-			wpa_printf(MSG_DEBUG,
+			wpa_printf(MSG_WARNING,
 				   "ACS: Channel %d: not allowed as primary channel for 40 MHz bandwidth",
 				   chan->chan);
 			continue;
@@ -904,7 +904,7 @@ acs_find_ideal_chan_mode(struct hostapd_
 			if (hostapd_get_oper_chwidth(iface->conf) ==
 			    CONF_OPER_CHWIDTH_80MHZ &&
 			    !acs_usable_bw_chan(chan, ACS_BW80)) {
-				wpa_printf(MSG_DEBUG,
+				wpa_printf(MSG_WARNING,
 					   "ACS: Channel %d: not allowed as primary channel for 80 MHz bandwidth",
 					   chan->chan);
 				continue;
@@ -913,7 +913,7 @@ acs_find_ideal_chan_mode(struct hostapd_
 			if (hostapd_get_oper_chwidth(iface->conf) ==
 			    CONF_OPER_CHWIDTH_160MHZ &&
 			    !acs_usable_bw_chan(chan, ACS_BW160)) {
-				wpa_printf(MSG_DEBUG,
+				wpa_printf(MSG_WARNING,
 					   "ACS: Channel %d: not allowed as primary channel for 160 MHz bandwidth",
 					   chan->chan);
 				continue;
@@ -943,7 +943,7 @@ acs_find_ideal_chan_mode(struct hostapd_
 				break;
 
 			if (!chan_bw_allowed(adj_chan, bw, 1, 0)) {
-				wpa_printf(MSG_DEBUG,
+				wpa_printf(MSG_WARNING,
 					   "ACS: PRI Channel %d: secondary channel %d BW %u is not supported",
 					   chan->chan, adj_chan->chan, bw);
 				break;
@@ -962,7 +962,7 @@ acs_find_ideal_chan_mode(struct hostapd_
 		}
 
 		if (j != n_chans) {
-			wpa_printf(MSG_DEBUG, "ACS: Channel %d: not enough bandwidth",
+			wpa_printf(MSG_WARNING, "ACS: Channel %d: not enough bandwidth",
 				   chan->chan);
 			continue;
 		}
@@ -1041,11 +1041,11 @@ acs_find_ideal_chan_mode(struct hostapd_
 
 		if (bias) {
 			factor *= bias->bias;
-			wpa_printf(MSG_DEBUG,
+			wpa_printf(MSG_WARNING,
 				   "ACS:  * channel %d: total interference = %Lg (%f bias)",
 				   chan->chan, factor, bias->bias);
 		} else {
-			wpa_printf(MSG_DEBUG,
+			wpa_printf(MSG_WARNING,
 				   "ACS:  * channel %d: total interference = %Lg",
 				   chan->chan, factor);
 		}
@@ -1124,13 +1124,16 @@ acs_find_ideal_chan(struct hostapd_iface
 		}
 	}
 
+	wpa_printf(MSG_WARNING, "ACS(%s) n_chans=%d 80211n=%d 80211ac=%d 80211xx=%d secondary_channel=%d", iface->phy, n_chans, iface->conf->ieee80211n, iface->conf->ieee80211ac, iface->conf->ieee80211ax, iface->conf->secondary_channel);
 	bw = num_chan_to_bw(n_chans);
 
 bw_selected:
+	// Register the bw for later use
+	iface->conf->bw=bw;
 	/* TODO: VHT/HE80+80. Update acs_adjust_center_freq() too. */
 
-	wpa_printf(MSG_DEBUG,
-		   "ACS: Survey analysis for selected bandwidth %d MHz", bw);
+	wpa_printf(MSG_WARNING,
+		   "ACS(%s): Survey analysis for selected bandwidth %d MHz", iface->phy, bw);
 
 	for (i = 0; i < iface->num_hw_features; i++) {
 		mode = &iface->hw_features[i];
@@ -1141,8 +1144,8 @@ bw_selected:
 	}
 
 	if (ideal_chan) {
-		wpa_printf(MSG_DEBUG, "ACS: Ideal channel is %d (%d MHz) with total interference factor of %Lg",
-			   ideal_chan->chan, ideal_chan->freq, ideal_factor);
+		wpa_printf(MSG_WARNING, "ACS(%s): Ideal channel is %d (%d MHz) with total interference factor of %Lg",
+			   iface->phy, ideal_chan->chan, ideal_chan->freq, ideal_factor);
 
 #ifdef CONFIG_IEEE80211BE
 		if (iface->conf->punct_acs_threshold)
@@ -1182,6 +1185,19 @@ static void acs_adjust_secondary(struct
 
 static void acs_adjust_center_freq(struct hostapd_iface *iface)
 {
+//	hostapd_ctrl_update_fixup( iface->conf );
+
+#ifdef CONFIG_IEEE80211AX
+	wpa_printf(MSG_WARNING, "ACS(%s): channel %u, secondary_channel %d, vht_center %u, he_center %u",
+		iface->phy, iface->conf->channel, iface->conf->secondary_channel,
+		iface->conf->vht_oper_centr_freq_seg0_idx, iface->conf->he_oper_centr_freq_seg0_idx);
+#else
+	wpa_printf(MSG_WARNING, "ACS(%s): channel %u, secondary_channel %d, vht_center %u",
+		iface->phy, iface->conf->channel, iface->conf->secondary_channel,
+		iface->conf->vht_oper_centr_freq_seg0_idx);
+#endif
+
+//#if 0
 	int center;
 
 	wpa_printf(MSG_DEBUG, "ACS: Adjusting center frequency");
@@ -1229,6 +1245,7 @@ static void acs_adjust_center_freq(struc
 	}
 
 	hostapd_set_oper_centr_freq_seg0_idx(iface->conf, center);
+//#endif
 }
 
 
@@ -1326,7 +1343,7 @@ static void acs_scan_complete(struct hos
 
 	iface->scan_cb = NULL;
 
-	wpa_printf(MSG_DEBUG, "ACS: Using survey based algorithm (acs_num_scans=%d)",
+	wpa_printf(MSG_WARNING, "ACS: Using survey based algorithm (acs_num_scans=%d)",
 		   iface->conf->acs_num_scans);
 
 	err = hostapd_drv_get_survey(iface->bss[0], 0);
@@ -1425,7 +1442,7 @@ static int acs_request_scan(struct hosta
 
 	iface->scan_cb = acs_scan_complete;
 
-	wpa_printf(MSG_DEBUG, "ACS: Scanning %d / %d",
+	wpa_printf(MSG_WARNING, "ACS: Scanning %d / %d",
 		   iface->acs_num_completed_scans + 1,
 		   iface->conf->acs_num_scans);
 
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1204,6 +1204,8 @@ struct hostapd_config {
 #define AIRTIME_MODE_MAX (__AIRTIME_MODE_MAX - 1)
 #endif /* CONFIG_AIRTIME_POLICY */
 
+	u32 bw;
+
 	int ieee80211be;
 #ifdef CONFIG_IEEE80211BE
 	enum oper_chan_width eht_oper_chwidth;
