--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -335,6 +335,7 @@ static int error_handler(struct sockaddr
 	struct nlattr *tb[NLMSGERR_ATTR_MAX + 1];
 	int *ret = arg;
 	int ack_len = sizeof(*nlh) + sizeof(int) + sizeof(*nlh);
+	static bool int_out_of_range_reported = false;
 
 	*ret = err->error;
 
@@ -354,6 +355,12 @@ static int error_handler(struct sockaddr
 	if (tb[NLMSGERR_ATTR_MSG]) {
 		len = strnlen((char *) nla_data(tb[NLMSGERR_ATTR_MSG]),
 			      nla_len(tb[NLMSGERR_ATTR_MSG]));
+		if (strcmp(nla_data(tb[NLMSGERR_ATTR_MSG]), "integer out of range") == 0) {
+			if (int_out_of_range_reported == false)
+				int_out_of_range_reported=true;
+			else
+				return NL_SKIP;
+		}
 		wpa_printf(MSG_ERROR, "nl80211: kernel reports: %*s",
 			   len, (char *) nla_data(tb[NLMSGERR_ATTR_MSG]));
 	}
