--- a/src/ap/acs.c
+++ b/src/ap/acs.c
@@ -390,6 +390,9 @@ acs_survey_interference_factor(struct fr
 		total -= survey->channel_time_tx;
 	}
 
+	if (!busy)
+		busy = 1;
+
 	/* TODO: figure out the best multiplier for noise floor base */
 	factor = pow(10, survey->nf / 5.0L) +
 		(total ? (busy / total) : 0) *
@@ -867,8 +870,8 @@ acs_find_ideal_chan_mode(struct hostapd_
 		    iface->conf->ieee80211n &&
 		    !acs_usable_bw_chan(chan, ACS_BW40)) {
 			wpa_printf(MSG_WARNING,
-				   "ACS: Channel %d: not allowed as primary channel for 40 MHz bandwidth",
-				   chan->chan);
+				   "ACS: Channel %d (freq %d): not allowed as primary channel for 40 MHz bandwidth",
+				   chan->chan, chan->freq);
 			continue;
 		}
 
@@ -877,15 +880,15 @@ acs_find_ideal_chan_mode(struct hostapd_
 		     iface->conf->ieee80211be)) {
 			if (bw == 80 && !acs_usable_bw_chan(chan, ACS_BW80)) {
 				wpa_printf(MSG_WARNING,
-					   "ACS: Channel %d: not allowed as primary channel for 80 MHz bandwidth",
-					   chan->chan);
+					   "ACS: Channel %d (freq %d): not allowed as primary channel for 80 MHz bandwidth",
+					   chan->chan, chan->freq);
 				continue;
 			}
 
 			if (bw == 160 && !acs_usable_bw_chan(chan, ACS_BW160)) {
 				wpa_printf(MSG_WARNING,
-					   "ACS: Channel %d: not allowed as primary channel for 160 MHz bandwidth",
-					   chan->chan);
+					   "ACS: Channel %d (freq %d): not allowed as primary channel for 160 MHz bandwidth",
+					   chan->chan, chan->freq);
 				continue;
 			}
 		}
@@ -896,8 +899,8 @@ acs_find_ideal_chan_mode(struct hostapd_
 			    !acs_usable_bw320_chan(iface, chan,
 						   &bw320_offset)) {
 				wpa_printf(MSG_WARNING,
-					   "ACS: Channel %d: not allowed as primary channel for 320 MHz bandwidth",
-					   chan->chan);
+					   "ACS: Channel %d (freq %d): not allowed as primary channel for 320 MHz bandwidth",
+					   chan->chan, chan->freq);
 				continue;
 			}
 		}
@@ -1046,12 +1049,12 @@ acs_find_ideal_chan_mode(struct hostapd_
 		if (bias) {
 			factor *= bias->bias;
 			wpa_printf(MSG_WARNING,
-				   "ACS:  * channel %d: total interference = %Lg (%f bias)",
-				   chan->chan, factor, bias->bias);
+				   "ACS:  * channel %d (%d): total interference = %Lg (%f bias)",
+				   chan->chan, chan->freq, factor, bias->bias);
 		} else {
 			wpa_printf(MSG_WARNING,
-				   "ACS:  * channel %d: total interference = %Lg",
-				   chan->chan, factor);
+				   "ACS:  * channel %d (%d): total interference = %Lg",
+				   chan->chan, chan->freq, factor);
 		}
 
 		if (acs_usable_chan(chan) &&
@@ -1129,7 +1132,9 @@ acs_find_ideal_chan(struct hostapd_iface
 		}
 	}
 
-	wpa_printf(MSG_WARNING, "ACS(%s) n_chans=%d 80211n=%d 80211ac=%d 80211xx=%d secondary_channel=%d", iface->phy, n_chans, iface->conf->ieee80211n, iface->conf->ieee80211ac, iface->conf->ieee80211ax, iface->conf->secondary_channel);
+	wpa_printf(MSG_WARNING, "ACS(%s) n_chans=%d 80211n=%d 80211ac=%d 80211ax=%d secondary_channel=%d",
+		iface->phy, n_chans, iface->conf->ieee80211n, iface->conf->ieee80211ac,
+		iface->conf->ieee80211ax, iface->conf->secondary_channel);
 	bw = num_chan_to_bw(n_chans);
 
 bw_selected:
