--- a/src/ap/acs.c
+++ b/src/ap/acs.c
@@ -871,6 +871,7 @@ acs_find_ideal_chan_mode(struct hostapd_
 	int bw320_offset = 0, ideal_bw320_offset = 0;
 	unsigned int k;
 	int secondary_channel = 1, freq_offset;
+	short last_chan = 0;
 
 	if (is_24ghz_mode(mode->mode))
 		secondary_channel = iface->conf->secondary_channel;
@@ -1104,15 +1105,19 @@ acs_find_ideal_chan_mode(struct hostapd_
 			bias = &tmp_bias;
 		}
 
-		if (bias) {
-			factor *= bias->bias;
-			wpa_printf(MSG_WARNING,
-				   "ACS:  * channel %d (%d): total interference = %Lg (%f bias)",
-				   chan->chan, chan->freq, factor, bias->bias);
-		} else {
-			wpa_printf(MSG_WARNING,
-				   "ACS:  * channel %d (%d): total interference = %Lg",
-				   chan->chan, chan->freq, factor);
+		/* Avoid displaying interference info more than once for the same channel */
+		if (chan->chan != last_chan) {
+			if (bias) {
+				factor *= bias->bias;
+				wpa_printf(MSG_WARNING,
+					   "ACS:  * channel %d (%d): total interference = %Lg (%f bias)",
+					   chan->chan, chan->freq, factor, bias->bias);
+			} else {
+				wpa_printf(MSG_WARNING,
+					   "ACS:  * channel %d (%d): total interference = %Lg",
+					   chan->chan, chan->freq, factor);
+			}
+			last_chan = chan->chan;
 		}
 
 		if (acs_usable_chan(chan) &&
