From 76e54b095fdd8ab65a562c28fba330afcfb519ec Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Wed, 5 Jul 2023 10:44:15 +0800
Subject: [PATCH] hostapd: mtk: Fix background channel overlapping operating
 channel issue

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 src/ap/dfs.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

--- a/src/ap/dfs.c
+++ b/src/ap/dfs.c
@@ -822,6 +822,20 @@ static int dfs_are_channels_overlapped(s
 }
 
 
+static void dfs_check_background_overlapped(struct hostapd_iface *iface)
+{
+	int width = hostapd_get_oper_chwidth(iface->conf);
+
+	if (!dfs_use_radar_background(iface))
+		return;
+
+	if (dfs_are_channels_overlapped(iface, iface->radar_background.freq,
+					width, iface->radar_background.centr_freq_seg0_idx,
+					iface->radar_background.centr_freq_seg1_idx))
+		iface->radar_background.channel = -1;
+}
+
+
 static unsigned int dfs_get_cac_time(struct hostapd_iface *iface,
 				     int start_chan_idx, int n_chans)
 {
@@ -1149,6 +1163,8 @@ static void hostapd_dfs_update_backgroun
 						  &oper_centr_freq_seg1_idx,
 						  &channel_type);
 	if (!channel ||
+	    channel->chan == iface->conf->channel ||
+	    channel->chan == iface->radar_background.channel ||
 	    hostapd_start_dfs_cac(iface, iface->conf->hw_mode,
 				  channel->freq, channel->chan,
 				  iface->conf->ieee80211n,
@@ -1393,6 +1409,7 @@ static int hostapd_dfs_start_channel_swi
 	hostapd_set_oper_centr_freq_seg1_idx(iface->conf,
 					     oper_centr_freq_seg1_idx);
 	err = 0;
+	dfs_check_background_overlapped(iface);
 
 	hostapd_setup_interface_complete(iface, err);
 	return err;
@@ -1520,6 +1537,7 @@ static int hostapd_dfs_start_channel_swi
 			hostapd_set_oper_centr_freq_seg1_idx(
 				iface->conf, oper_centr_freq_seg1_idx);
 
+			dfs_check_background_overlapped(iface);
 			hostapd_disable_iface(iface);
 			hostapd_enable_iface(iface);
 			return 0;
