--- a/include/iwinfo.h
+++ b/include/iwinfo.h
@@ -438,6 +438,8 @@ extern const struct iwinfo_ops madwifi_o
 extern const struct iwinfo_ops nl80211_ops;
 extern const struct iwinfo_ops wl_ops;
 
+int nl80211_get_htmodelistfilter(const char *ifname, int *buf, int band);
+
 #include "iwinfo/utils.h"
 
 #endif
--- a/iwinfo_cli.c
+++ b/iwinfo_cli.c
@@ -18,6 +18,7 @@
 
 #include <stdio.h>
 #include <glob.h>
+#include <assert.h>
 
 #include "iwinfo.h"
 
@@ -922,6 +923,33 @@ static void print_htmodelist(const struc
 	printf("\n");
 }
 
+static void print_htmodelistfilter(const char *ifname, const char *band_str)
+{
+	int i, band = 0, htmodes = 0;
+
+	if (band_str != NULL)
+	{
+		if (band_str[0] == '2')
+			band = IWINFO_BAND_24; 
+		else if (band_str[0] == '5')
+			band = IWINFO_BAND_5;
+		else if (band_str[0] == '6')
+			band = IWINFO_BAND_6;
+	}
+
+	if (nl80211_get_htmodelistfilter(ifname, &htmodes, band))
+	{
+		printf("No HT mode information available\n");
+		return;
+	}
+
+	for (i = 0; i < IWINFO_HTMODE_COUNT; i++)
+		if (htmodes & (1 << i))
+			printf("%s ", IWINFO_HTMODE_NAMES[i]);
+
+	printf("\n");
+}
+
 static void lookup_phy(const struct iwinfo_ops *iw, const char *section)
 {
 	char buf[IWINFO_BUFSIZE];
@@ -969,7 +997,7 @@ int main(int argc, char **argv)
 			"	iwinfo <device> freqlist\n"
 			"	iwinfo <device> assoclist\n"
 			"	iwinfo <device> countrylist\n"
-			"	iwinfo <device> htmodelist\n"
+			"	iwinfo <device> htmodelist [2|5|6]\n"
 			"	iwinfo <backend> phyname <section>\n"
 		);
 
@@ -1000,7 +1028,7 @@ int main(int argc, char **argv)
 		return 0;
 	}
 
-	if (argc > 3)
+	if (argc > 3 && argv[2][0] == 'p')
 	{
 		iw = iwinfo_backend_by_name(argv[1]);
 
@@ -1026,6 +1054,30 @@ int main(int argc, char **argv)
 				rv = 1;
 			}
 		}
+	}
+	else if (argc == 4 && argv[2][0] == 'h')
+	{
+		iw = iwinfo_backend(argv[1]);
+		assert(strcmp(iw->name, "nl80211") == 0);
+
+		if (!iw)
+		{
+			fprintf(stderr, "No such wireless device: %s\n", argv[1]);
+			rv = 1;
+		}
+		else
+		{
+			switch(argv[2][0])
+			{
+			case 'h':
+				print_htmodelistfilter(argv[1], argv[3]);
+				break;
+
+			default:
+				fprintf(stderr, "Unknown command: %s\n", argv[i]);
+				rv = 1;
+			}
+		}
 	}
 	else
 	{
--- a/iwinfo_nl80211.c
+++ b/iwinfo_nl80211.c
@@ -3719,6 +3719,35 @@ out:
 	return -1;
 }
 
+int nl80211_get_htmodelistfilter(const char *ifname, int *buf, int band)
+{
+	struct nl80211_msg_conveyor *cv;
+	struct nl80211_modes m = {};
+	uint32_t features = nl80211_get_protocol_features(ifname);
+	int flags;
+
+	m.radio_band = band;
+
+	flags = features & NL80211_PROTOCOL_FEATURE_SPLIT_WIPHY_DUMP ? NLM_F_DUMP : 0;
+	cv = nl80211_msg(ifname, NL80211_CMD_GET_WIPHY, flags);
+	if (!cv)
+		goto out;
+
+	NLA_PUT_FLAG(cv->msg, NL80211_ATTR_SPLIT_WIPHY_DUMP);
+	if (nl80211_send(cv, nl80211_get_modelist_cb, &m))
+		goto nla_put_failure;
+
+	nl80211_eval_modelist(&m);
+
+	*buf = m.ht;
+
+	return 0;
+
+nla_put_failure:
+	nl80211_free(cv);
+out:
+	return -1;
+}
 
 static int nl80211_get_ifcomb_cb(struct nl_msg *msg, void *arg)
 {
