Index: backports-5.15.58-1/net/wireless/util.c
===================================================================
--- backports-5.15.58-1.orig/net/wireless/util.c
+++ backports-5.15.58-1/net/wireless/util.c
@@ -557,7 +557,7 @@ int ieee80211_data_to_8023_exthdr(struct
 		return -1;
 
 	hdrlen = ieee80211_hdrlen(hdr->frame_control) + data_offset;
-	if (skb->len < hdrlen + 8)
+	if (iftype == NL80211_IFTYPE_MESH_POINT && skb->len < hdrlen + 8)
 		return -1;
 
 	/* convert IEEE 802.11 header + possible LLC headers into Ethernet
Index: backports-5.15.58-1/net/mac80211/rx.c
===================================================================
--- backports-5.15.58-1.orig/net/mac80211/rx.c
+++ backports-5.15.58-1/net/mac80211/rx.c
@@ -4647,7 +4647,7 @@ static bool ieee80211_invoke_fast_rx(str
 
 	if (!(status->rx_flags & IEEE80211_RX_AMSDU)) {
 		if (!pskb_may_pull(skb, snap_offs + sizeof(*payload)))
-			goto drop;
+			return false;
 
 		payload = (void *)(skb->data + snap_offs);
 
