Index: backports-5.15-rc6-1/net/mac80211/iface.c
===================================================================
--- backports-5.15-rc6-1.orig/net/mac80211/iface.c
+++ backports-5.15-rc6-1/net/mac80211/iface.c
@@ -842,13 +842,17 @@ static const struct net_device_ops ieee8
 
 };
 
-static bool ieee80211_iftype_supports_hdr_offload(enum nl80211_iftype iftype)
+static bool ieee80211_iftype_supports_hdr_offload(enum nl80211_iftype iftype, int decap)
 {
 	switch (iftype) {
 	/* P2P GO and client are mapped to AP/STATION types */
 	case NL80211_IFTYPE_AP:
-	case NL80211_IFTYPE_STATION:
 		return true;
+	case NL80211_IFTYPE_STATION:
+		if (decap)
+			return false;
+		else
+			return true;
 	default:
 		return false;
 	}
@@ -862,7 +866,7 @@ static bool ieee80211_set_sdata_offload_
 	flags = sdata->vif.offload_flags;
 
 	if (ieee80211_hw_check(&local->hw, SUPPORTS_TX_ENCAP_OFFLOAD) &&
-	    ieee80211_iftype_supports_hdr_offload(sdata->vif.type)) {
+	    ieee80211_iftype_supports_hdr_offload(sdata->vif.type, 0)) {
 		flags |= IEEE80211_OFFLOAD_ENCAP_ENABLED;
 
 		if (!ieee80211_hw_check(&local->hw, SUPPORTS_TX_FRAG) &&
@@ -876,7 +880,7 @@ static bool ieee80211_set_sdata_offload_
 	}
 
 	if (ieee80211_hw_check(&local->hw, SUPPORTS_RX_DECAP_OFFLOAD) &&
-	    ieee80211_iftype_supports_hdr_offload(sdata->vif.type)) {
+	    ieee80211_iftype_supports_hdr_offload(sdata->vif.type, 1)) {
 		flags |= IEEE80211_OFFLOAD_DECAP_ENABLED;
 
 		if (local->monitors &&
@@ -908,7 +912,7 @@ static void ieee80211_set_vif_encap_ops(
 	}
 
 	if (!ieee80211_hw_check(&local->hw, SUPPORTS_TX_ENCAP_OFFLOAD) ||
-	    !ieee80211_iftype_supports_hdr_offload(bss->vif.type))
+	    !ieee80211_iftype_supports_hdr_offload(bss->vif.type, 0))
 		return;
 
 	enabled = bss->vif.offload_flags & IEEE80211_OFFLOAD_ENCAP_ENABLED;
