Index: mt76-2021-04-04-f8c6c7cb/mt7615/init.c
===================================================================
--- mt76-2021-04-04-f8c6c7cb.orig/mt7615/init.c
+++ mt76-2021-04-04-f8c6c7cb/mt7615/init.c
@@ -118,7 +118,7 @@ mt7615_mac_init(struct mt7615_dev *dev)
 
 	mt76_wr(dev, MT_DMA_DCR0,
 		FIELD_PREP(MT_DMA_DCR0_MAX_RX_LEN, 3072) |
-		MT_DMA_DCR0_RX_VEC_DROP | MT_DMA_DCR0_DAMSDU_EN |
+		MT_DMA_DCR0_RX_VEC_DROP |
 		MT_DMA_DCR0_RX_HDR_TRANS_EN);
 	/* disable TDLS filtering */
 	mt76_clear(dev, MT_WF_PFCR, MT_WF_PFCR_TDLS_EN);
