From: Scott Iveland <scott.iveland@smartrg.com>
Date: Fri, 30 Jul 2021 13:18:48 -0700
Subject: mediatek: add option to read rf data from a firmware file

Signed-off-by: Scott Iveland <scott.iveland@smartrg.com>
--
eeprom.c |  103 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 102 insertions(+), 1 deletion(-)

===================================================================
--- a/eeprom.c
+++ b/eeprom.c
@@ -9,6 +9,107 @@
 #include <linux/etherdevice.h>
 #include "mt76.h"
 
+int mt76_get_of_eeprom_file(struct mt76_dev *dev, void *eep, int offset, int len)
+{
+#if defined(CONFIG_OF)
+	struct device_node *np = dev->dev->of_node;
+	const __be32 *list;
+	const char *part;
+	phandle phandle;
+	int size;
+	int ret=0;
+	int retlen;
+
+	char path[64]="";
+	struct file *fp;
+	loff_t pos=0;
+	struct inode *inode = NULL;
+	loff_t f_size;
+
+	if (!np)
+		return -ENOENT;
+
+	list = of_get_property(np, "mediatek,mtd-eeprom", &size);
+	if (!list)
+		return -ENOENT;
+
+	phandle = be32_to_cpup(list++);
+	if (!phandle)
+		return -ENOENT;
+
+	np = of_find_node_by_phandle(phandle);
+	if (!np)
+		return -EINVAL;
+
+	part = of_get_property(np, "label", NULL);
+	if (!part)
+		part = np->name;
+
+	dev_dbg(dev->dev,"Looking for /lib/firmware/mediatek/rf.bin");
+	retlen = snprintf(path,sizeof(path),"/lib/firmware/mediatek/rf.bin");
+	if(retlen<0)
+		return -EINVAL;
+
+	fp = filp_open(path, O_RDONLY, 0);
+	if (IS_ERR(fp)) {
+		dev_warn(dev->dev,"Open eeprom file failed: %s\n",path);
+		return -ENOENT;
+	}
+
+	inode = file_inode(fp);
+	if ((!S_ISREG(inode->i_mode) && !S_ISBLK(inode->i_mode))) {
+		printk(KERN_ALERT "invalid file type: %s\n", path);
+		return -ENOENT;
+	}
+	f_size = i_size_read(inode->i_mapping->host);
+	if (f_size < 0)
+	{
+		printk(KERN_ALERT "failed getting size of %s size:%lld \n",path,f_size);
+		return -ENOENT;
+	}
+
+	if (size <= sizeof(*list)) {
+		ret = -EINVAL;
+		goto out_put_node;
+	}
+
+	offset += be32_to_cpup(list);
+	pos = offset;
+
+	dev_dbg(dev->dev,"Reading file: len %d, pos %lld \n",len,pos);
+	retlen = kernel_read(fp, dev->eeprom.data, len, &pos);
+	if (retlen != len) {
+		ret = -EINVAL;
+		dev_warn(dev->dev,"Load eeprom ERR, count %d byte (len:%d)\n",ret,len);
+		goto out_put_node;
+	}
+
+	if (of_property_read_bool(dev->dev->of_node, "big-endian")) {
+		u8 *data = (u8 *)eep;
+		int i;
+
+		/* convert eeprom data in Little Endian */
+		for (i = 0; i < round_down(len, 2); i += 2)
+			put_unaligned_le16(get_unaligned_be16(&data[i]),
+					   &data[i]);
+	}
+
+#ifdef CONFIG_NL80211_TESTMODE
+	dev->test_mtd.name = devm_kstrdup(dev->dev, part, GFP_KERNEL);
+	dev->test_mtd.offset = offset;
+#endif
+
+	dev_dbg(dev->dev,"Load eeprom from rf.bin OK, count %d, pos %lld \n",retlen,pos);
+
+out_put_node:
+	of_node_put(np);
+	filp_close(fp, 0);
+	return ret;
+#else
+	return -ENOENT;
+#endif
+}
+
 int mt76_get_of_eeprom(struct mt76_dev *dev, void *eep, int offset, int len)
 {
 #if defined(CONFIG_OF) && defined(CONFIG_MTD)
@@ -338,6 +439,6 @@ mt76_eeprom_init(struct mt76_dev *dev, i
 	if (!dev->eeprom.data)
 		return -ENOMEM;
 
-	return !mt76_get_of_eeprom(dev, dev->eeprom.data, 0, len);
+	return (!mt76_get_of_eeprom_file(dev, dev->eeprom.data, 0, len)) || (!mt76_get_of_eeprom(dev, dev->eeprom.data, 0, len));
 }
 EXPORT_SYMBOL_GPL(mt76_eeprom_init);
