--- a/mt7915/mac.c
+++ b/mt7915/mac.c
@@ -749,10 +749,17 @@ mt7915_mac_fill_rx(struct mt7915_dev *de
 			return -EINVAL;
 		hdr_trans = false;
 	} else {
+		int pad_start = 0;
+
 		skb_pull(skb, hdr_gap);
-		if (!hdr_trans && status->amsdu) {
-			memmove(skb->data + 2, skb->data,
-				ieee80211_get_hdrlen_from_skb(skb));
+		if (!hdr_trans && status->amsdu)
+			pad_start = ieee80211_get_hdrlen_from_skb(skb);
+		else if (hdr_trans &&
+				get_unaligned_be16(skb->data + 12) == ETH_P_8021Q)
+			pad_start = 16;
+
+		if (pad_start) {
+			memmove(skb->data + 2, skb->data, pad_start);
 			skb_pull(skb, 2);
 		}
 	}
