From 84449ab5c15f96d3212e467ee6f66e6827219ea2 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Mon, 28 Aug 2023 10:59:23 +0800
Subject: [PATCH] wifi: mt76: mt7915: check vif type before reporting cca and
 csa

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
Signed-off-by: rany <ranygh@riseup.net>
---
 mt7915/mcu.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/mt7915/mcu.c
+++ b/mt7915/mcu.c
@@ -225,7 +225,7 @@ int mt7915_mcu_wa_cmd(struct mt7915_dev
 static void
 mt7915_mcu_csa_finish(void *priv, u8 *mac, struct ieee80211_vif *vif)
 {
-	if (vif->bss_conf.csa_active)
+	if (vif->bss_conf.csa_active && vif->type != NL80211_IFTYPE_STATION)
 		ieee80211_csa_finish(vif);
 }
 
@@ -332,7 +332,7 @@ mt7915_mcu_rx_log_message(struct mt7915_
 static void
 mt7915_mcu_cca_finish(void *priv, u8 *mac, struct ieee80211_vif *vif)
 {
-	if (!vif->bss_conf.color_change_active)
+	if (!vif->bss_conf.color_change_active || vif->type == NL80211_IFTYPE_STATION)
 		return;
 
 	ieee80211_color_change_finish(vif);
