--- a/mt7615/init.c
+++ b/mt7615/init.c
@@ -431,9 +431,15 @@ mt7615_cap_dbdc_enable(struct mt7615_dev
 static void
 mt7615_cap_dbdc_disable(struct mt7615_dev *dev)
 {
+#if 0
 	dev->mphy.sband_5g.sband.vht_cap.cap |=
 			IEEE80211_VHT_CAP_SHORT_GI_160 |
 			IEEE80211_VHT_CAP_SUPP_CHAN_WIDTH_160_80PLUS80MHZ;
+#else
+	dev->mphy.sband_5g.sband.vht_cap.cap &=
+			~(IEEE80211_VHT_CAP_SHORT_GI_160 |
+			  IEEE80211_VHT_CAP_SUPP_CHAN_WIDTH_160_80PLUS80MHZ);
+#endif
 	dev->mphy.antenna_mask = dev->chainmask;
 	dev->mphy.chainmask = dev->chainmask;
 	dev->mphy.hw->wiphy->available_antennas_rx = dev->chainmask;
