--- a/mt7915/init.c
+++ b/mt7915/init.c
@@ -393,10 +393,12 @@ mt7915_init_wiphy(struct ieee80211_hw *h
 				IEEE80211_VHT_CAP_MAX_MPDU_LENGTH_7991 |
 				IEEE80211_VHT_CAP_MAX_A_MPDU_LENGTH_EXPONENT_MASK;
 
+#if 0
 			if (!dev->dbdc_support)
 				phy->mt76->sband_5g.sband.vht_cap.cap |=
 					IEEE80211_VHT_CAP_SHORT_GI_160 |
 					IEEE80211_VHT_CAP_SUPP_CHAN_WIDTH_160_80PLUS80MHZ;
+#endif
 		} else {
 			phy->mt76->sband_5g.sband.vht_cap.cap |=
 				IEEE80211_VHT_CAP_MAX_MPDU_LENGTH_11454 |
@@ -877,11 +879,8 @@ mt7915_init_he_caps(struct mt7915_phy *p
 
 	if (!is_mt7915(&dev->mt76))
 		nss_160 = nss;
-	else if (!dev->dbdc_support)
-		/* Can do 1/2 of NSS streams in 160Mhz mode for mt7915 */
-		nss_160 = nss / 2;
 	else
-		/* Can't do 160MHz with mt7915 dbdc */
+		/* don't allow 160MHz with mt7915 */
 		nss_160 = 0;
 
 	for (i = 0; i < 8; i++) {
