Index: mt76-2021-10-20-b649678c/mt7615/pci_init.c
===================================================================
--- mt76-2021-10-20-b649678c.orig/mt7615/pci_init.c
+++ mt76-2021-10-20-b649678c/mt7615/pci_init.c
@@ -28,8 +28,6 @@ static void mt7615_pci_init_work(struct
 		return;
 
 	mt7615_init_work(dev);
-	if (dev->dbdc_support)
-		mt7615_register_ext_phy(dev);
 }
 
 static int mt7615_init_hardware(struct mt7615_dev *dev)
@@ -160,6 +158,9 @@ int mt7615_register_device(struct mt7615
 	mt7615_init_txpower(dev, &dev->mphy.sband_2g.sband);
 	mt7615_init_txpower(dev, &dev->mphy.sband_5g.sband);
 
+	if (dev->dbdc_support)
+		mt7615_register_ext_phy(dev);
+
 	return mt7615_init_debugfs(dev);
 }
 
