From 56916977daae3317294a905950cdc604d048bddb Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Wed, 2 Aug 2023 11:26:14 +0800
Subject: [PATCH] wifi: mt76: mt7915: fix tx emission issue during bootup after
 applying pre-cal

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Reviewed-by: Shayne Chen <shayne.chen@mediatek.com>
Reviewed-by: Evelyn Tsai <evelyn.tsai@mediatek.com>
---
 mt7915/mcu.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/mt7915/mcu.c
+++ b/mt7915/mcu.c
@@ -2771,7 +2771,8 @@ int mt7915_mcu_set_chan_info(struct mt79
 
 	if (dev->mt76.hw->conf.flags & IEEE80211_CONF_MONITOR)
 		req.switch_reason = CH_SWITCH_NORMAL;
-	else if (phy->mt76->hw->conf.flags & IEEE80211_CONF_OFFCHANNEL)
+	else if (phy->mt76->hw->conf.flags & IEEE80211_CONF_OFFCHANNEL ||
+		 phy->mt76->hw->conf.flags & IEEE80211_CONF_IDLE)
 		req.switch_reason = CH_SWITCH_SCAN_BYPASS_DPD;
 	else if (cfg80211_chandef_valid(chandef) &&
 		 !cfg80211_reg_can_beacon(phy->mt76->hw->wiphy, chandef,
