--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -658,6 +658,9 @@ mt7996_mcu_update_rate(struct rate_info
 	case MT_PHY_TYPE_EHT_TRIG:
 	case MT_PHY_TYPE_EHT_MU:
 		tmp_rate.mcs = mcs & GENMASK(3, 0);
+		if (tmp_rate.mcs > 13)
+			tmp_rate.mcs = 13;
+
 		if (tmp_rate.mcs > 15 || gi > NL80211_RATE_INFO_EHT_GI_3_2)
 			return -EINVAL;
 
@@ -700,11 +703,10 @@ mt7996_mcu_update_rate(struct rate_info
 }
 
 static int
-mt7996_mcu_update_trx_rates(struct mt76_wcid *wcid, struct all_sta_trx_rate *mcu_rate)
+mt7996_mcu_update_trx_rates(struct mt7996_dev *dev, struct mt76_wcid *wcid, struct all_sta_trx_rate *mcu_rate)
 {
-	struct mt7996_sta_link *msta_link = container_of(wcid, struct mt7996_sta_link, wcid);
-	struct mt76_dev *dev = &msta_link->sta->vif->dev->mt76;
-	struct mt76_phy *phy = mt76_dev_phy(dev, wcid->phy_idx);
+	struct mt76_dev *mdev = &dev->mt76;
+	struct mt76_phy *mphy = mt76_dev_phy(mdev, wcid->phy_idx);
 	struct ieee80211_supported_band *sband = NULL;
 	bool cck;
 	int ret;
@@ -717,14 +719,14 @@ mt7996_mcu_update_trx_rates(struct mt76_
 		cck = true;
 		fallthrough;
 	case MT_PHY_TYPE_OFDM:
-		if (phy->chandef.chan->band == NL80211_BAND_2GHZ) {
-			sband = &phy->sband_2g.sband;
+		if (mphy->chandef.chan->band == NL80211_BAND_2GHZ) {
+			sband = &mphy->sband_2g.sband;
 			if (!cck)
 				mcu_rate->tx_mcs += 4;
-		} else if (phy->chandef.chan->band == NL80211_BAND_5GHZ)
-			sband = &phy->sband_5g.sband;
+		} else if (mphy->chandef.chan->band == NL80211_BAND_5GHZ)
+			sband = &mphy->sband_5g.sband;
 		else
-			sband = &phy->sband_6g.sband;
+			sband = &mphy->sband_6g.sband;
 		break;
 	case MT_PHY_TYPE_HT:
 	case MT_PHY_TYPE_HT_GF:
@@ -738,8 +740,13 @@ mt7996_mcu_update_trx_rates(struct mt76_
 				     mcu_rate->tx_bw, mcu_rate->tx_mcs,
 				     mcu_rate->tx_nss, mcu_rate->tx_stbc,
 				     mcu_rate->tx_gi);
-	if (ret)
+	if (ret) {
+		dev_err(mdev->dev, "%s: failed to update TX rate mode=%u bw=%u mcs=%u nss=%u stbc=%u gi=%u",
+				__func__, mcu_rate->tx_mode, mcu_rate->tx_bw, mcu_rate->tx_mcs,
+				mcu_rate->tx_nss, mcu_rate->tx_stbc, mcu_rate->tx_gi);
+
 		return ret;
+	}
 
 	/* RX rate */
 	cck = false;
@@ -749,14 +756,14 @@ mt7996_mcu_update_trx_rates(struct mt76_
 		cck = true;
 		fallthrough;
 	case MT_PHY_TYPE_OFDM:
-		if (phy->chandef.chan->band == NL80211_BAND_2GHZ)
-			sband = &phy->sband_2g.sband;
-		else if (phy->chandef.chan->band == NL80211_BAND_5GHZ)
-			sband = &phy->sband_5g.sband;
+		if (mphy->chandef.chan->band == NL80211_BAND_2GHZ)
+			sband = &mphy->sband_2g.sband;
+		else if (mphy->chandef.chan->band == NL80211_BAND_5GHZ)
+			sband = &mphy->sband_5g.sband;
 		else
-			sband = &phy->sband_6g.sband;
+			sband = &mphy->sband_6g.sband;
 
-		mcu_rate->rx_rate = mt76_get_rate(dev, sband, mcu_rate->rx_rate, cck);
+		mcu_rate->rx_rate = mt76_get_rate(mdev, sband, mcu_rate->rx_rate, cck);
 		break;
 	default:
 		break;
@@ -766,7 +773,16 @@ mt7996_mcu_update_trx_rates(struct mt76_
 				     mcu_rate->rx_bw, mcu_rate->rx_rate,
 				     mcu_rate->rx_nsts + 1, mcu_rate->rx_stbc,
 				     mcu_rate->rx_gi);
-	return ret;
+
+	if (ret) {
+		dev_err(mdev->dev, "%s: failed to update RX rate mode=%u bw=%u mcs=%u nss=%u stbc=%u gi=%u",
+				__func__, mcu_rate->rx_mode, mcu_rate->rx_bw, mcu_rate->rx_rate,
+				mcu_rate->rx_nsts + 1, mcu_rate->rx_stbc, mcu_rate->rx_gi);
+
+		return ret;
+	}
+
+	return 0;
 }
 
 static void
@@ -781,6 +797,7 @@ mt7996_mcu_rx_all_sta_info_event(struct
 
 	for (i = 0; i < le16_to_cpu(res->sta_num); i++) {
 		u8 ac;
+		bool v1;
 		u16 wlan_idx;
 		bool clear;
 		struct mt76_wcid *wcid;
@@ -795,21 +812,33 @@ mt7996_mcu_rx_all_sta_info_event(struct
 			if (!wcid)
 				break;
 
-			if (mt7996_mcu_update_trx_rates(wcid, &res->rate[i]))
-				dev_err(dev->mt76.dev, "Failed to update TX/RX rates.\n");
+			if (mt7996_mcu_update_trx_rates(dev, wcid, &res->rate[i]))
+				dev_dbg(dev->mt76.dev, "failed to update TX/RX rates");
 			break;
 		case UNI_ALL_STA_TXRX_ADM_STAT:
-			wlan_idx = le16_to_cpu(res->adm_stat[i].wlan_idx);
+			v1 = le16_to_cpu(res->len) == UNI_EVENT_SIZE_ADM_STAT_V1;
+			if (v1)
+				wlan_idx = le16_to_cpu(res->adm_stat_v1[i].wlan_idx);
+			else
+				wlan_idx = le16_to_cpu(res->adm_stat_v2[i].wlan_idx);
 			wcid = mt76_wcid_ptr(dev, wlan_idx);
 
 			if (!wcid)
 				break;
 
 			for (ac = 0; ac < IEEE80211_NUM_ACS; ac++) {
-				wcid->stats.tx_bytes +=
-					le32_to_cpu(res->adm_stat[i].tx_bytes[ac]);
-				wcid->stats.rx_bytes +=
-					le32_to_cpu(res->adm_stat[i].rx_bytes[ac]);
+				u8 lmac_ac = mt76_connac_lmac_mapping(ac);
+				if (v1) {
+					wcid->stats.tx_bytes +=
+						le32_to_cpu(res->adm_stat_v1[i].tx_bytes[lmac_ac]);
+					wcid->stats.rx_bytes +=
+						le32_to_cpu(res->adm_stat_v1[i].rx_bytes[lmac_ac]);
+				} else {
+					wcid->stats.tx_bytes +=
+						le32_to_cpu(res->adm_stat_v2[i].tx_bytes[lmac_ac]);
+					wcid->stats.rx_bytes +=
+						le32_to_cpu(res->adm_stat_v2[i].rx_bytes[lmac_ac]);
+				}
 			}
 			break;
 		case UNI_ALL_STA_TXRX_MSDU_COUNT:
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -1831,14 +1831,8 @@ static void mt7996_sta_statistics(struct
 				  struct ieee80211_sta *sta,
 				  struct station_info *sinfo)
 {
-	struct mt7996_dev *dev = mt7996_hw_dev(hw);
 	struct mt7996_sta *msta = (struct mt7996_sta *)sta->drv_priv;
-	struct mt7996_sta_link *msta_link;
-
-	mutex_lock(&dev->mt76.mutex);
-	msta_link = mt76_dereference(msta->link[msta->deflink_id], &dev->mt76);
-	if (!msta_link)
-		goto out;
+	struct mt7996_sta_link *msta_link = &msta->deflink;
 
 	sinfo->txrate = msta_link->wcid.rate;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_TX_BITRATE);
@@ -1869,9 +1863,6 @@ static void mt7996_sta_statistics(struct
 
 	sinfo->rx_packets = msta_link->wcid.stats.rx_packets;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_RX_PACKETS);
-
-out:
-	mutex_unlock(&dev->mt76.mutex);
 }
 
 static void mt7996_link_rate_ctrl_update(void *data,
--- a/mt7996/mcu.h
+++ b/mt7996/mcu.h
@@ -263,6 +263,8 @@ struct all_sta_trx_rate {
 	u8 __rsv2;
 } __packed;
 
+#define UNI_EVENT_SIZE_ADM_STAT_V1 1452
+
 struct mt7996_mcu_all_sta_info_event {
 	u8 rsv[4];
 	__le16 tag;
@@ -279,7 +281,15 @@ struct mt7996_mcu_all_sta_info_event {
 			u8 rsv[2];
 			__le32 tx_bytes[IEEE80211_NUM_ACS];
 			__le32 rx_bytes[IEEE80211_NUM_ACS];
-		} __packed, adm_stat);
+		} __packed, adm_stat_v1);
+
+		DECLARE_FLEX_ARRAY(struct {
+			__le16 wlan_idx;
+			u8 rsv[2];
+			__le32 tx_bytes[IEEE80211_NUM_ACS];
+			__le32 rx_bytes[IEEE80211_NUM_ACS];
+			__le32 tx_bytes_failed[IEEE80211_NUM_ACS];
+		} __packed, adm_stat_v2);
 
 		DECLARE_FLEX_ARRAY(struct {
 			__le16 wlan_idx;
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -50,6 +50,9 @@ static struct mt76_wcid *mt7996_rx_get_w
 		break;
 	}
 
+	if (!msta_link)
+		return wcid;
+
 	return &msta_link->wcid;
 }
 
--- a/mt7996/init.c
+++ b/mt7996/init.c
@@ -598,11 +598,6 @@ mt7996_mac_init_band(struct mt7996_dev *
 	set = FIELD_PREP(MT_WTBLOFF_RSCR_RCPI_MODE, 0) |
 	      FIELD_PREP(MT_WTBLOFF_RSCR_RCPI_PARAM, 0x3);
 	mt76_rmw(dev, MT_WTBLOFF_RSCR(band), mask, set);
-
-	/* MT_TXD5_TX_STATUS_HOST (MPDU format) has higher priority than
-	 * MT_AGG_ACR_PPDU_TXS2H (PPDU format) even though ACR bit is set.
-	 */
-	mt76_set(dev, MT_AGG_ACR4(band), MT_AGG_ACR_PPDU_TXS2H);
 }
 
 static void mt7996_mac_init_basic_rates(struct mt7996_dev *dev)
