--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -70,13 +70,13 @@ int mt7996_run(struct ieee80211_hw *hw)
 	if (ret)
 		goto out;
 
-	ret = mt7996_mcu_set_scs(phy, SCS_ENABLE);
+	ret = mt7996_mcu_set_scs(phy, SCS_DISABLE);
 	if (ret)
 		goto out;
 
 #ifdef CONFIG_MTK_DEBUG
-	phy->sr_enable = true;
-	phy->enhanced_sr_enable = true;
+	phy->sr_enable = false;
+	phy->enhanced_sr_enable = false;
 
 	dev_info(dev->mt76.dev, "%s: sku enable = %d, path enable = %d\n", __func__,
 		dev->dbg.sku_disable ? 0 : phy->sku_limit_en, dev->dbg.sku_disable ? 0 : phy->sku_path_en);
