--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -3983,6 +3983,7 @@ int mt7996_mcu_get_temperature(struct mt
 	} __packed * res;
 	struct sk_buff *skb;
 	int ret;
+	u32 val;
 
 	ret = mt76_mcu_send_and_get_msg(&phy->dev->mt76, MCU_WM_UNI_CMD(THERMAL),
 					&req, sizeof(req), true, &skb);
@@ -3990,8 +3991,10 @@ int mt7996_mcu_get_temperature(struct mt
 		return ret;
 
 	res = (void *)skb->data;
+	val = le32_to_cpu(res->temperature);
+	dev_kfree_skb(skb);
 
-	return le32_to_cpu(res->temperature);
+	return val;
 }
 
 int mt7996_mcu_set_thermal_throttling(struct mt7996_phy *phy, u8 state)
