--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -848,6 +848,7 @@ mt7996_mcu_rx_all_sta_info_event(struct
 	if (sta_num > (skb->len - sizeof(*res)) / elem_size)
 		sta_num = (skb->len - sizeof(*res)) / elem_size;
 
+	rcu_read_lock();
 	for (i = 0; i < sta_num; i++) {
 		u8 ac;
 		bool v1;
@@ -966,6 +967,7 @@ mt7996_mcu_rx_all_sta_info_event(struct
 			break;
 		}
 	}
+	rcu_read_unlock();
 }
 
 static void
