--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -738,6 +738,7 @@ mt7996_mcu_rx_all_sta_info_event(struct
 
 	res = (struct mt7996_mcu_all_sta_info_event *)skb->data;
 
+	rcu_read_lock();
 	for (i = 0; i < le16_to_cpu(res->sta_num); i++) {
 		u8 ac;
 		bool v1;
@@ -904,6 +905,7 @@ mt7996_mcu_rx_all_sta_info_event(struct
 			break;
 		}
 	}
+	rcu_read_unlock();
 }
 
 static void
@@ -5267,6 +5269,8 @@ int mt7996_mcu_get_per_sta_info(struct m
 		for (i = 0; i < sta_num; ++i) {
 			wlan_idx = le16_to_cpu(res->rssi[i].wlan_idx);
 			wcid = rcu_dereference(dev->wcid[wlan_idx]);
+			if (!wcid || !wcid->sta || wcid->sta_disabled)
+				continue;
 			msta_link = container_of(wcid, struct mt7996_sta_link, wcid);
 			if (msta_link) {
 				struct mt76_phy *phy = dev->phys[wcid->phy_idx];
@@ -5285,6 +5289,8 @@ int mt7996_mcu_get_per_sta_info(struct m
 		for (i = 0; i < sta_num; ++i) {
 			wlan_idx = le16_to_cpu(res->snr[i].wlan_idx);
 			wcid = rcu_dereference(dev->wcid[wlan_idx]);
+			if (!wcid || !wcid->sta || wcid->sta_disabled)
+				continue;
 			msta_link = container_of(wcid, struct mt7996_sta_link, wcid);
 			if (msta_link)
 				memcpy(msta_link->chain_ack_snr, res->snr[i].val,
@@ -5295,7 +5301,7 @@ int mt7996_mcu_get_per_sta_info(struct m
 		for (i = 0; i < sta_num; ++i) {
 			wlan_idx = le16_to_cpu(res->msdu_cnt[i].wlan_idx);
 			wcid = rcu_dereference(dev->wcid[wlan_idx]);
-			if (wcid) {
+			if (wcid && wcid->sta && !wcid->sta_disabled) {
 				u32 retries = le32_to_cpu(res->msdu_cnt[i].tx_retries),
 				    drops = le32_to_cpu(res->msdu_cnt[i].tx_drops);
 
