---
 mt7996/eeprom.c |  9 +++++----
 mt7996/mcu.c    | 18 +-----------------
 2 files changed, 6 insertions(+), 21 deletions(-)

--- a/mt7996/eeprom.c
+++ b/mt7996/eeprom.c
@@ -192,13 +192,13 @@ static int mt7996_eeprom_load_flash(stru
 
 		if (mt7996_eeprom_load_default(dev))
 			return 0;
-
-		if (mt7996_check_eeprom(dev))
-			return 0;
 	} else {
 		ret = mt76_eeprom_init(&dev->mt76, MT7996_EEPROM_SIZE);
 	}
 
+	if (mt7996_check_eeprom(dev))
+		return 0;
+
 	return ret;
 }
 
@@ -226,11 +226,12 @@ static int mt7996_eeprom_load(struct mt7
 		if (free_block_num >= 59)
 			return -EINVAL;
 
+		memset(dev->mt76.eeprom.data, 0, MT7996_EEPROM_SIZE);
 		/* read eeprom data from efuse */
 		block_num = DIV_ROUND_UP(MT7996_EEPROM_SIZE, eeprom_blk_size);
 		for (i = 0; i < block_num; i++) {
 			ret = mt7996_mcu_get_eeprom(dev, i * eeprom_blk_size, NULL);
-			if (ret < 0)
+			if (ret && ret != -EINVAL)
 				return ret;
 		}
 		dev->eeprom_mode = EFUSE_MODE;
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -3569,7 +3569,7 @@ int mt7996_mcu_set_chan_info(struct mt79
 				 &req, sizeof(req), true);
 }
 
-static int mt7996_mcu_set_eeprom_flash(struct mt7996_dev *dev)
+int mt7996_mcu_set_eeprom(struct mt7996_dev *dev)
 {
 #define MAX_PAGE_IDX_MASK	GENMASK(7, 5)
 #define PAGE_IDX_MASK		GENMASK(4, 2)
@@ -3614,22 +3614,6 @@ static int mt7996_mcu_set_eeprom_flash(s
 	return 0;
 }
 
-int mt7996_mcu_set_eeprom(struct mt7996_dev *dev)
-{
-	struct mt7996_mcu_eeprom req = {
-		.tag = cpu_to_le16(UNI_EFUSE_BUFFER_MODE),
-		.len = cpu_to_le16(sizeof(req) - 4),
-		.buffer_mode = EE_MODE_EFUSE,
-		.format = EE_FORMAT_WHOLE
-	};
-
-	if (dev->flash_mode)
-		return mt7996_mcu_set_eeprom_flash(dev);
-
-	return mt76_mcu_send_msg(&dev->mt76, MCU_WM_UNI_CMD(EFUSE_CTRL),
-				 &req, sizeof(req), true);
-}
-
 int mt7996_mcu_get_eeprom(struct mt7996_dev *dev, u32 offset, u8 *read_buf)
 {
 	struct mt7996_mcu_eeprom_info req = {
