From 14c5941e4049b07c9f3b67075dd80a97dd785a65 Mon Sep 17 00:00:00 2001
From: TomLiu <tomml.liu@mediatek.com>
Date: Tue, 13 Aug 2024 22:30:20 +0000
Subject: [PATCH] [mt76]mt7996 add airtime filter for airtime spike

---
 mcu.c | 30 ++++++++++++++++++++----------
 1 file changed, 20 insertions(+), 10 deletions(-)

--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -4041,16 +4041,26 @@ int mt7996_mcu_get_chan_mib_info(struct
 		goto out;
 
 #define __res_u64(s) le64_to_cpu(res[s].data)
-	state->cc_tx += __res_u64(IDX_TX_TIME) - state_ts->cc_tx;
-	state->cc_bss_rx += __res_u64(IDX_RX_TIME) - state_ts->cc_bss_rx;
-	state->cc_rx += __res_u64(IDX_RX_TIME) +
-			__res_u64(IDX_OBSS_AIRTIME) -
-			state_ts->cc_rx;
-	state->cc_busy += __res_u64(IDX_TX_TIME) +
-			  __res_u64(IDX_RX_TIME) +
-			  __res_u64(IDX_OBSS_AIRTIME) +
-			  __res_u64(IDX_NON_WIFI_TIME) -
-			  state_ts->cc_busy;
+	if (__res_u64(IDX_TX_TIME) - state_ts->cc_tx < 150000) {
+		state->cc_tx += __res_u64(IDX_TX_TIME) - state_ts->cc_tx;
+	}
+	if (__res_u64(IDX_RX_TIME) - state_ts->cc_bss_rx < 150000) {
+		state->cc_bss_rx += __res_u64(IDX_RX_TIME) - state_ts->cc_bss_rx;
+	}
+	if (__res_u64(IDX_RX_TIME) + __res_u64(IDX_OBSS_AIRTIME) - 
+			state_ts->cc_rx < 150000) {
+		state->cc_rx += __res_u64(IDX_RX_TIME) +
+				__res_u64(IDX_OBSS_AIRTIME) -
+				state_ts->cc_rx;
+	}
+	if (__res_u64(IDX_TX_TIME) + __res_u64(IDX_RX_TIME) + __res_u64(IDX_OBSS_AIRTIME) +
+                          __res_u64(IDX_NON_WIFI_TIME) - state_ts->cc_busy < 150000) {
+		state->cc_busy += __res_u64(IDX_TX_TIME) +
+				  __res_u64(IDX_RX_TIME) +
+				  __res_u64(IDX_OBSS_AIRTIME) +
+				  __res_u64(IDX_NON_WIFI_TIME) -
+				  state_ts->cc_busy;
+	}
 out:
 	state_ts->cc_tx = __res_u64(IDX_TX_TIME);
 	state_ts->cc_bss_rx = __res_u64(IDX_RX_TIME);
