--- a/mt7915/mac.c
+++ b/mt7915/mac.c
@@ -1356,10 +1356,10 @@ out:
 static void
 mt7915_mac_full_reset(struct mt7915_dev *dev)
 {
-	struct mt76_phy *ext_phy;
 	int i;
-
-	ext_phy = dev->mt76.phys[MT_BAND1];
+	struct mt76_phy *ext_phy = dev->mt76.phys[MT_BAND1];
+	struct mt7915_phy *phy = &dev->phy;
+	struct mt7915_phy *phy2 = ext_phy ? ext_phy->priv : NULL;
 
 	dev->recovery.hw_full_reset = true;
 
@@ -1389,6 +1389,18 @@ mt7915_mac_full_reset(struct mt7915_dev
 		list_del_init(dev->mt76.sta_poll_list.next);
 	spin_unlock_bh(&dev->mt76.sta_poll_lock);
 
+	spin_lock_bh(&phy->stats_lock);
+	while (!list_empty(&phy->stats_list))
+		list_del_init(phy->stats_list.next);
+	spin_unlock_bh(&phy->stats_lock);
+
+	if (phy2) {
+		spin_lock_bh(&phy2->stats_lock);
+		while (!list_empty(&phy2->stats_list))
+			list_del_init(phy2->stats_list.next);
+		spin_unlock_bh(&phy2->stats_lock);
+	}
+
 	memset(dev->mt76.wcid_mask, 0, sizeof(dev->mt76.wcid_mask));
 	dev->mt76.vif_mask = 0;
 
