--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -1737,6 +1737,7 @@ mt7996_mac_full_reset(struct mt7996_dev
 	phy3 = mt7996_phy3(dev);
 	dev->recovery.hw_full_reset = true;
 
+	set_bit(MT76_RESET, &dev->mphy.state);
 	wake_up(&dev->mt76.mcu.wait);
 	ieee80211_stop_queues(mt76_hw(dev));
 	if (phy2)
@@ -1757,36 +1758,29 @@ mt7996_mac_full_reset(struct mt7996_dev
 		if (!mt7996_mac_restart(dev))
 			break;
 	}
-	mutex_unlock(&dev->mt76.mutex);
 
 	if (i == 10)
 		dev_err(dev->mt76.dev, "chip full reset failed\n");
 
-	ieee80211_restart_hw(mt76_hw(dev));
+	dev->phy.omac_mask = 0;
 	if (phy2)
-		ieee80211_restart_hw(phy2->mt76->hw);
+		phy2->omac_mask = 0;
 	if (phy3)
-		ieee80211_restart_hw(phy3->mt76->hw);
+		phy3->omac_mask = 0;
 
-	ieee80211_wake_queues(mt76_hw(dev));
-	if (phy2)
-		ieee80211_wake_queues(phy2->mt76->hw);
-	if (phy3)
-		ieee80211_wake_queues(phy3->mt76->hw);
+	mt76_reset_device(&dev->mt76);
+
+	INIT_LIST_HEAD(&dev->sta_rc_list);
+	INIT_LIST_HEAD(&dev->twt_list);
+
+	i = mt76_wcid_alloc(dev->mt76.wcid_mask, MT7996_WTBL_STA);
+	dev->mt76.global_wcid.idx = i;
 
 	dev->recovery.hw_full_reset = false;
-	ieee80211_queue_delayed_work(mt76_hw(dev),
-				     &dev->mphy.mac_work,
-				     MT7996_WATCHDOG_TIME);
-	if (phy2)
-		ieee80211_queue_delayed_work(phy2->mt76->hw,
-					     &phy2->mt76->mac_work,
-					     MT7996_WATCHDOG_TIME);
-	if (phy3)
-		ieee80211_queue_delayed_work(phy3->mt76->hw,
-					     &phy3->mt76->mac_work,
-					     MT7996_WATCHDOG_TIME);
-	ieee80211_queue_delayed_work(mt76_hw(dev), &dev->scs_work, HZ);
+
+	mutex_unlock(&dev->mt76.mutex);
+
+	ieee80211_restart_hw(mt76_hw(dev));
 }
 
 void mt7996_mac_reset_work(struct work_struct *work)
@@ -1859,6 +1853,10 @@ void mt7996_mac_reset_work(struct work_s
 		set_bit(MT76_RESET, &phy3->mt76->state);
 		cancel_delayed_work_sync(&phy3->mt76->mac_work);
 	}
+	cancel_delayed_work_sync(&dev->scs_work);
+
+	mutex_lock(&dev->mt76.mutex);
+
 	mt76_worker_disable(&dev->mt76.tx_worker);
 
 	dev_info(dev->mt76.dev,"%s L1 SER disable tx_work done.",
@@ -1873,8 +1871,6 @@ void mt7996_mac_reset_work(struct work_s
 	}
 	napi_disable(&dev->mt76.tx_napi);
 
-	mutex_lock(&dev->mt76.mutex);
-
 	dev_info(dev->mt76.dev,"%s L1 SER napi disable done.",
 		 wiphy_name(dev->mt76.hw->wiphy));
 
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -2291,6 +2291,29 @@ mt7996_change_vif_links(struct ieee80211
 	return 0;
 }
 
+static void
+mt7996_reconfig_complete(struct ieee80211_hw *hw,
+			 enum ieee80211_reconfig_type reconfig_type)
+{
+	struct mt7996_dev *dev = mt7996_hw_dev(hw);
+	struct mt76_dev *mdev = &dev->mt76;
+
+	ieee80211_wake_queues(hw);
+	ieee80211_queue_delayed_work(hw, &mdev->phy.mac_work,
+				     MT7996_WATCHDOG_TIME);
+
+	if (mdev->phys[MT_BAND1])
+		ieee80211_queue_delayed_work(hw, &mdev->phys[MT_BAND1]->mac_work,
+						 MT7996_WATCHDOG_TIME);
+
+	if (mdev->phys[MT_BAND2])
+		ieee80211_queue_delayed_work(hw, &mdev->phys[MT_BAND2]->mac_work,
+						 MT7996_WATCHDOG_TIME);
+
+	//if (phy->scs_ctrl.scs_enable)
+	//	ieee80211_queue_delayed_work(mt76_hw(dev), &dev->scs_work, HZ);
+}
+
 const struct ieee80211_ops mt7996_ops = {
 	.add_chanctx = mt76_add_chanctx,
 	.remove_chanctx = mt76_remove_chanctx,
@@ -2349,4 +2372,5 @@ const struct ieee80211_ops mt7996_ops =
 #endif
 	.change_vif_links = mt7996_change_vif_links,
 	.change_sta_links = mt7996_mac_sta_change_links,
+	.reconfig_complete = mt7996_reconfig_complete,
 };
